web_app/contract_tools/ekubo_research.py

```diff
@@ -0,0 +1,203 @@
+
+Key Concepts:
+------------
+1. Flash Accounting: Ekubo's internal token balance accounting system.
+2. Free Flash Loans: No fees for borrowing within same transaction.
+3. Atomic Execution: All operations must complete in single transaction.
+
+Implementation Flow:
+------------------
+
+API Endpoints(https://mainnet-api.ekubo.org/openapi.json) and Data Structures
+"""
+
+from dataclasses import dataclass
+from typing import List, Dict
+from decimal import Decimal
+import httpx
+
+@dataclass
+class PoolKey:
+    """Ekubo pool identification structure"""
+    token0: str
+    token1: str
+    fee: int
+    tick_spacing: int
+    extension: str
+
+@dataclass
+class SwapRoute:
+    """Structure for defining swap path"""
+    pool_key: PoolKey
+    sqrt_ratio_limit: int
+    skip_ahead: int
+
+@dataclass
+class TokenAmount:
+    """Token amount with direction"""
+    token: str
+    amount: int
+    is_positive: bool
+
+class EkuboFlashLoan:
+    """
+    Main class for handling Ekubo Flash Loan operations
+    
+    Key Features:
+    - No upfront fees required
+    - Must repay within same transaction
+    - Uses flash accounting for internal balance management
+    """
+
+    def __init__(self, rpc_url: str, core_contract: str):
+        self.rpc_url = rpc_url
+        self.core_contract = core_contract
+
+    async def get_profitable_route(self, input_token: str, amount: Decimal) -> List[SwapRoute]:
+        """
+        Find profitable arbitrage route using Ekubo API
+        
+        Example API call:
+        GET https://mainnet-api.ekubo.org/quote/{amount}/{input_token}/{output_token}
+        """
+        base_url = "https://mainnet-api.ekubo.org/quote"
+        amount_wei = int(amount * Decimal(10**18))
+
+        async with httpx.AsyncClient() as client:
+            # Get quote for potential arbitrage route
+            response = await client.get(
+                f"{base_url}/{amount_wei}/{input_token}/{input_token}"
+            )
+            quote_data = response.json()
+
+            # Parse route from response
+            return self._parse_route(quote_data['route'])
+
+    def prepare_flash_loan_tx(
+        self,
+        route: List[SwapRoute],
+        borrow_amount: TokenAmount
+    ) -> Dict:
+        """
+        Prepare transaction for flash loan arbitrage
+        
+        Transaction flow:
+        1. Borrow tokens via flash accounting (no explicit borrow call needed)
+        2. Execute swaps through provided route
+        3. Verify profit
+        4. Repay loan automatically from swap proceeds
+        """
+        return {
+            'contract_address': self.core_contract,
+            'entry_point_selector': 'multi_route_swap',
+            'calldata': [
+                # Calldata format:
+                # 1. Number of route nodes
+                # 2. For each node: pool_key, sqrt_ratio_limit, skip_ahead
+                # 3. Token amount with direction
+                len(route),
```

Comment:
I haven't seen any contract endpoints with such a signature, so it would be great to add what contract you investigated and refer to as core: address or docs URL is enough.